### PR TITLE
[STK-118][FEAT] - Enhance user balances using Multicall contract

### PR DESCRIPTION
**Pull request by ElRodrigote**
### Description

* Batch fetch ERC-20 token balances from the user's connected wallet using the Multicall smart contract
* Descending sort those balances

### Preview
image

### How to test the changes
1) Pull this branch
2) Run the project locally
3) Go to the Stackly homepage, connect your wallet, and then click any of the two "Select token" buttons

### CODE REVIEW

File: packages/app/contexts/TokenListContext.tsx [hunk 1]
```diff
@@ -29,8 +37,15 @@ const TOKEN_LIST_BY_CHAIN_URL: { [chainId: number]: string } = {
   100: "https://tokens.honeyswap.org/",
 };
 
+const provider = new ethers.providers.JsonRpcProvider(
+  "https://rpc.gnosischain.com"
+);
```
Comment (ElRodrigote) [inline packages/app/contexts/TokenListContext.tsx:42]: This is the thing I hate the most in the code I wrote, but I'm not 100% sure how to programmatically set this RPC provider here.

Tried using the [default provider](https://docs.ethers.org/v5/api/providers/#providers-getDefaultProvider) but it kept failing :c

File: packages/app/contexts/TokenListContext.tsx [hunk 2]
```diff
@@ -97,9 +122,49 @@ export const TokenListProvider = ({ children }: PropsWithChildren) => {
...
+   * all the ERC-20 token balances for the connected wallet
+   * and the supported token list, then we desc. sort them
+   */
+  useEffect(() => {
+    if (address) {
+      provider
+        .call({
+          to: MULTICALL_ADDRESS,
+          data: multicallInterface.encodeFunctionData("aggregate", [callArray]),
+        })
```
Comment (berteotti) [inline packages/app/contexts/TokenListContext.tsx:138]: Maybe we should be using https://wagmi.sh/core/actions/multicall here

Comment (berteotti) [reply packages/app/contexts/TokenListContext.tsx:138]: This will probably let us free from `provider`

**Review by berteotti — COMMENTED** [review]
@ElRodrigote answering your comment on the following review :)

File: packages/app/contexts/TokenListContext.tsx (hunk 2 above)
Comment (ElRodrigote) [reply packages/app/contexts/TokenListContext.tsx:138]: Never used it, will try it out and see the results :D

Comment (ElRodrigote) [reply packages/app/contexts/TokenListContext.tsx]: Made this change after updating `viem` and `wagmi` deps, thanks a lot for the off-line guidance messages :D

File: packages/app/contexts/TokenListContext.tsx [hunk 3]
```diff
@@ -97,9 +105,54 @@ export const TokenListProvider = ({ children }: PropsWithChildren) => {
...
+              chainId: token.chainId,
+            })),
+            allowFailure: true,
+          });
+
+          const listWithBalances = tokenList
+            .map((token, index) => ({
+              ...token,
+              balance: formatUnits(
+                batchFetchdata[index]?.result as bigint,
```
Comment (Diogomartf) [inline packages/app/contexts/TokenListContext.tsx:133]: Just wondering if it's safe to assume the index will match the correct `token` on `batchFetchdata` 🤔

**Comment by Diogomartf** [comment]
Not a big deal but when we disconnect wallet we are keeping the balances.

image

**Review by Diogomartf — DISMISSED** [review]
The balances look snappy and code is looking good. Good job 👍

File: packages/app/contexts/TokenListContext.tsx (hunk 1 above)
Comment (Diogomartf) [reply packages/app/contexts/TokenListContext.tsx]: related to this, I this this code from @berteotti mint/gatekeeping logic PR can be used here as well

https://github.com/SwaprHQ/stackly-ui/pull/75/files#diff-b0ad06a11e66920646084475d9e320a6a38b6ec59ea4dcd4dd02f622c14e7ddaR32-R52

File: packages/app/contexts/TokenListContext.tsx (hunk 3 above)
Comment (ElRodrigote) [reply packages/app/contexts/TokenListContext.tsx:133]: It is an array based on the `tokenList`, it _should_ have the same indexing

File: packages/app/components/TokenIcon.tsx [hunk 4]
```diff
@@ -18,15 +18,9 @@ export const TokenIcon = ({ className, size, token }: TokenIconProps) => {
   const { getTokenFromList, getTokenLogoURL } = useTokenListContext();
   const [isImgBroken, setIsImgBroken] = useState<Boolean>(false);
 
-  const getTokenMethodsUndefined = !getTokenFromList || !getTokenLogoURL;
-  const invalidAddress = !isAddress(token.address);
-  const noTokenOnTheList =
-    !getTokenMethodsUndefined && !getTokenFromList(token.address);
-
```
Comment (berteotti) [inline packages/app/components/TokenIcon.tsx:25]: Less code, nice!

**Review by Diogomartf — DISMISSED** [review]
good job!

**Comment by Diogomartf** [comment]
just look at the CI build, it's failing

**Comment by ElRodrigote** [comment]
> Collaborator

Yes! Dunno was failing before, afterwards I got the pipelines failing due to merge conflicts.
Everything seems to be solved now :D

**Review by berteotti — APPROVED** [review]
Good job Rorry!

**Pull request merged**